### PR TITLE
Fixing `vec_normal` function

### PR DESCRIPTION
**Pull request by PNieck**
## Description

In this PR behavior of function `vec_normal` was changed. Now function returns perpendicular vector, when paraller vectors are passed to it. Moreover there are tests for this function.

## Motivation and Context

Better interface for vectors library.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
https://github.com/phoenix-pilot/phoenix-pilot-core/issues/117
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: (add PR link here).In this PR
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: algebra/vec.c [hunk 1]
```diff
@@ -80,10 +80,30 @@ float vec_len(const vec_t *A)
...
+		len = vec_len(C);
+
+		if (len == 0) {
+			v.z = 1;
+
+			vec_cross(&longerV, &v, C);
+			len = vec_len(C);
+
+			if (len == 0) {
+				vec_times(C, 0);
```
Comment (niewim19) [inline algebra/vec.c:100]: You were right with multiplication being useless. Lets just hand-type there zeroes.

File: algebra/tests/vec/various.c [hunk 2]
```diff
@@ -21,6 +21,8 @@
 	TEST_ASSERT_EQUAL_FLOAT_MESSAGE((expected).y, (actual).y, "Different `y` part of vectors"); \
 	TEST_ASSERT_EQUAL_FLOAT_MESSAGE((expected).z, (actual).z, "Different `z` part of vectors");
 
+#define DELTA 5e-6
```
Comment (niewim19) [inline algebra/tests/vec/various.c:24]: Add comment regarding this macro, please.

File: algebra/tests/vec/various.c [hunk 3]
```diff
@@ -502,3 +506,133 @@ TEST_GROUP_RUNNER(group_vec_cross)
...
+
+TEST(group_vec_normal, vec_normal_std)
+{
+	vec_t A = V1;
+	vec_t B = V2;
+	vec_t C;
+
+	vec_normal(&A, &B, &C);
+
+	TEST_ASSERT_FLOAT_WITHIN(DELTA, 0.0, vec_dot(&A, &C));
```
Comment (niewim19) [inline algebra/tests/vec/various.c:537]: How about new macro `TEST_ASSERT_PERPENDICULAR_VEC()`?

File: algebra/tests/vec/various.c [hunk 4]
```diff
@@ -502,3 +506,133 @@ TEST_GROUP_RUNNER(group_vec_cross)
...
+
+
+TEST_TEAR_DOWN(group_vec_normal)
+{
+}
+
+
+TEST(group_vec_normal, vec_normal_std)
+{
+	vec_t A = V1;
```
Comment (niewim19) [inline algebra/tests/vec/various.c:531]: as discussed - please use small letters in new PR-s

File: algebra/tests/vec/various.c [hunk 5]
```diff
@@ -502,3 +506,133 @@ TEST_GROUP_RUNNER(group_vec_cross)
...
+
+	vec_normal(&B, &A, &C);
+
+	TEST_ASSERT_FLOAT_WITHIN(DELTA, 0.0, vec_dot(&A, &C));
+	TEST_ASSERT_FLOAT_WITHIN(DELTA, 0.0, vec_dot(&B, &C));
+	TEST_ASSERT_EQUAL_FLOAT(1.0, vec_len(&C));
+}
+
+
+TEST(group_vec_normal, vec_normal_oneZeroVector)
```
Comment (niewim19) [inline algebra/tests/vec/various.c:598]: Please add mathematical themed comment regarding this edge case

File: algebra/tests/vec/various.c [hunk 6]
```diff
@@ -502,3 +506,133 @@ TEST_GROUP_RUNNER(group_vec_cross)
...
+
+	vec_normal(&B, &A, &C);
+
+	TEST_ASSERT_FLOAT_WITHIN(DELTA, 0.0, vec_dot(&A, &C));
+	TEST_ASSERT_FLOAT_WITHIN(DELTA, 0.0, vec_dot(&B, &C));
+	TEST_ASSERT_EQUAL_FLOAT(1.0, vec_len(&C));
+}
+
+
+TEST(group_vec_normal, vec_normal_bothZeroVectors)
```
Comment (niewim19) [inline algebra/tests/vec/various.c:618]: Please add mathematical themed comment regarding this corner case

**Review by niewim19 — CHANGES_REQUESTED** [review]

File: algebra/tests/vec/various.c (hunk 3 above)
Comment (PNieck) [reply algebra/tests/vec/various.c:537]: Yes, I also thought about it

File: algebra/vec.c [hunk 7]
```diff
@@ -79,11 +79,35 @@ float vec_len(const vec_t *A)
...
-	float len;
+	float lenSquared;
+	vec_t v = { .x = 1.0f, .y = 0.0f, .y = 0.0f, .z = 0.0f };
+	const vec_t *longerV;
 
 	vec_cross(A, B, C);
-	len = vec_len(C);
-	(len != 0) ? vec_times(C, 1.F / len) : vec_times(C, 0);
+	lenSquared = vec_dot(C, C);
+	if (lenSquared == 0) {
```
Comment (niewim19) [inline algebra/vec.c:88]: How about dropping the far nesting of `if`s and use multiple returns? Something like:
 - do a cross
 - if `len != 0` return
 - do different cross
 - if `len != 0` return
 - do different cross
 - final if and returns

**Review by niewim19 — CHANGES_REQUESTED** [review]

File: algebra/vec.c [hunk 8]
```diff
@@ -79,11 +79,35 @@ float vec_len(const vec_t *A)
 
 void vec_normal(const vec_t *A, const vec_t *B, vec_t *C)
 {
-	float len;
+	float lenSquared;
+	vec_t v = { .x = 1.0f, .y = 0.0f, .y = 0.0f, .z = 0.0f };
+	const vec_t *longerV;
```
[the remark this answers is not in the document]
Comment (PNieck) [reply algebra/vec.c:84]: I cannot do that, because I am changing `v` in this function

File: algebra/vec.c [hunk 9]
```diff
@@ -79,11 +79,41 @@ float vec_len(const vec_t *A)
...
-	len = vec_len(C);
-	(len != 0) ? vec_times(C, 1.F / len) : vec_times(C, 0);
+	lenSquared = vec_dot(C, C);
+
+	if (lenSquared != 0) {
+		vec_times(C, 1.F / sqrtf(lenSquared));
+		return;
+	}
+
+	longerV = (vec_dot(A, A) > vec_dot(B, B)) ? A : B;
```
Comment (niewim19) [inline algebra/vec.c:94]: Please, add comments after each, non-final return regarding what is the next scenario

File: algebra/tests/vec/various.c [hunk 10]
```diff
@@ -634,3 +646,133 @@ TEST_GROUP_RUNNER(group_vec_len)
...
+
+	TEST_ASSERT_PERPENDICULAR_VEC(DELTA, a, c);
+	TEST_ASSERT_PERPENDICULAR_VEC(DELTA, b, c);
+	TEST_ASSERT_UNIT_VEC(c);
+
+	vec_normal(&a, &b, &c);
+
+	TEST_ASSERT_PERPENDICULAR_VEC(DELTA, a, c);
+	TEST_ASSERT_PERPENDICULAR_VEC(DELTA, b, c);
+	TEST_ASSERT_UNIT_VEC(c);
```
Comment (niewim19) [inline algebra/tests/vec/various.c:755]: Isn't it supposed to have different arguments order? I can`t find the difference between this part, and above

**Review by niewim19 — CHANGES_REQUESTED** [review]

**Review by niewim19 — DISMISSED** [review]
LGTM
